flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSetJsonSerializer.java

```diff
@@ -0,0 +1,89 @@
+import org.apache.flink.formats.json.JsonRowDataSerializationSchema;
+import org.apache.flink.table.catalog.Column;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonGenerator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.JsonSerializer;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.SerializerProvider;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.table.gateway.api.results.ResultSet.FIELD_NAME_COLUMN_INFOS;
+import static org.apache.flink.table.gateway.api.results.ResultSet.FIELD_NAME_DATA;
+import static org.apache.flink.table.gateway.api.results.ResultSet.FIELD_NAME_NEXT_TOKEN;
+import static org.apache.flink.table.gateway.api.results.ResultSet.FIELD_NAME_RESULT_TYPE;
+
+/** Json serializer for {@link ResultSet}. */
+public class ResultSetJsonSerializer extends JsonSerializer<ResultSet> {
+
+    @Override
+    public void serialize(
+            ResultSet resultSet, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+            throws IOException {
+        jsonGenerator.writeStartObject();
+
+        // Serialize result type
+        serializerProvider.defaultSerializeField(
+                FIELD_NAME_RESULT_TYPE, resultSet.getResultType(), jsonGenerator);
+
+        // Serialize next token
+        serializerProvider.defaultSerializeField(
+                FIELD_NAME_NEXT_TOKEN, resultSet.getNextToken(), jsonGenerator);
+
+        // Serialize column infos
+        List<Column> columns = resultSet.getResultSchema().getColumns();
+        List<ColumnInfo> columnInfos = new ArrayList<>();
+        for (Column column : columns) {
+            columnInfos.add(
+                    ColumnInfo.create(column.getName(), column.getDataType().getLogicalType()));
+        }
+        serializerProvider.defaultSerializeField(
+                FIELD_NAME_COLUMN_INFOS, columnInfos, jsonGenerator);
+
+        // Serialize RowData
+        JsonRowDataSerializationSchema serializationSchema =
+                new JsonRowDataSerializationSchema(
+                        (RowType)
+                                resultSet
+                                        .getResultSchema()
+                                        .toPhysicalRowDataType()
+                                        .getLogicalType(),
+                        TimestampFormat.ISO_8601,
+                        JsonFormatOptions.MapNullKeyMode.LITERAL,
+                        "null",
+                        true);
+        List<String> rowDataStrings = new ArrayList<>();
+        for (RowData row : resultSet.getData()) {
+            byte[] rowByte = serializationSchema.serialize(row);
+            rowDataStrings.add(new String(rowByte, StandardCharsets.UTF_8));
+        }
```

Comment:
These doesn't align with the FLIP.

Reply:
Fixed. 
Now `data` contains a list of Json Objects. Each of Json Object contains a RowKind and a list of JsonNode.